### ISSUE TITLE
Support unique indexes on `TEXT` fields without a prefix length (MariaDB compatibility)

### ISSUE THREAD

**Issue by xEinfachJoshua** [dolthub/dolt#7040]
hello guys i found a other bug while importing a sql file
failed to run import job: failed to import sql: error on line 4191 for query
--
-- Indizes für die Tabelle codem_mechanic_boss
--
ALTER TABLE codem_mechanic_boss
ADD UNIQUE KEY name (name) USING HASH: blob/text column 'name' used in key specification without a key length
[Bild[](url)](https://cdn.discordapp.com/attachments/788576421370920971/1176952134516887654/image.png?ex=6570bd17&is=655e4817&hm=280f2bdef019ef5cf4fbe16530595e0802fecf19f6167b8bf36144d3c2cb2781&)

**Comment by nicktobey** [comment]
From the MariaDB docs:

```
Starting with [MariaDB 10.4](https://mariadb.com/kb/en/what-is-mariadb-104/), a unique index can be created on a TEXT column.

Internally, this uses hash indexing to quickly check the values and if a hash collision is found, the actual stored values are compared in order to retain the uniqueness.
```

This is different from MySQL, which does not allow indexes on TEXT columns unless there's a prefix length.

We should be able to support this: TEXT columns are stored in the db as a content-addressed hash, which makes it unsuited for iteration, but easy to do point lookups or uniqueness checks. (I'm assuming that even in MariaDB this index can't be used for range lookups, just point lookups and validating uniqueness.)

**Comment by fulghum** [comment]
Thanks for reporting this one @xEinfachJoshua. In general, I'd say our highest goal is making Dolt be a drop-in replacement for MySQL, and after that, I'd optimistically prioritize compatibility with MariaDB, since MariaDB does expand on MySQL support in several places (e.g. application time periods are a cool MariaDB feature not in MySQL) and we want MariaDB customers to be able to move to Dolt easily, too. 

I'm going to start digging into this one to better understand the scope and I'll follow up with an update.

**Comment by fulghum** [comment]
I'm getting back to this after the holiday break and I've got a rough prototype that shows the work needed to support this...

The trickiest bit so far is that Dolt's current index code assumes that if a BLOB or TEXT type is used in an index, then it will **always** have a prefix length specified. Because of this assumption, address encoded types like BLOB and TEXT are automatically converted to non-address encoded types in the index's key schema so that the raw string can be stored in the index file. However, for this feature, we don't want to store the content and instead want to store a hash of the content; reusing the content-encoded address seems convenient for that. I've hacked up a code path that does that and the unique index seems to be working with that content hash, but it's specific to my test, and not a general solution. To generalize this, we need to perform a different schema mapping for unique indexes that contain BLOB/TEXT columns that didn't get a prefix specified and it looks like it may be a bit messy to wire all of that context up where it's needed. One other twist with this is that if there is a collision on the content hash, we need to do additional work to compare the two real values and validate if there really is a duplicate value. 

After getting the index key schema working correctly, the other part needed for this change is to ensure that these indexes aren't used for range scans on the columns that are represented as content hashes.

**Comment by fulghum** [comment]
I've been making some good progress on this one, here's a quick summary of the work breakdown and status of each piece:
- Disable prefix length validation for BLOB/TEXT fields in unique indexes ✅
- Add a SQL system var to opt-in to strict MySQL compatibility (`strict_mysql_compatibility`) ✅
- Store content hash in index 🟡
	- This is working for Dolt's index implementation, but I think there is a small amount of work remaining in the go-mysql-server layer. Tests are passing, but I need to dig in and make sure they are running correctly. 
- Prevent indexes with content-hashed fields from being used in range scans 🟡
	- This is implemented for range scans, but there's a somewhat similar issue with using the index for join planning that needs to be investigated. 
- Error messaging: `ErrUniqueKeyViolation` 🔴
	- The error message for a unique constraint violation on a content-hashed field currently includes the content hash, not the real content. This error is coming from the GMS layer, so may be a little tricky to wire this through. 
	- This could potentially be a follow up, but without this, the unique constraint error message really isn't useful at all. 
- On hash collision, compare the real content 🔴
	- Without this, there is a risk of false positives – if two different values have the same content hash, they will trigger a unique constraint error. 
	- This could potentially be done as a follow up. There is no risk of violating the unique index and introducing duplicate values (since the content hash would catch that), but there is a risk that a valid value cannot be inserted if it has a duplicate hash. 

Other than that, there are a few more edge cases I want to test, such as an index with one content-hashed TEXT/BLOB field and one non-content-hashed TEXT/BLOB field.

**Comment by fulghum** [comment]
Support for this feature has been merged in and will go out in our next Dolt release. Once it's released, I'll update this issue again and mark as resolved. 

Note that the initial version of this feature still has some limitations. See https://github.com/dolthub/dolt/issues/7132 for more details. We will prioritize that work based on customer feedback, so feel free to comment on that new GitHub issue and let us know if you are running into problems with this feature and need any of those enhancements.

**Comment by fulghum** [comment]
@xEinfachJoshua – This feature has now been released in [Dolt version 1.29.3](https://github.com/dolthub/dolt/releases/tag/v1.29.3). 

Thank you for taking the time to report this! 🙏  It turned out to be a really interesting feature to work on and sparked some good discussions around compatibility with MySQL extensions in MariaDB. Note that there are still [a few limitations](https://github.com/dolthub/dolt/issues/7132) in this initial version. We'll be happy to prioritize those as customer need them. 

I hope you'll give this new Dolt version a shot and let us know if you have more feedback or hit more issues!
